File: 01_multi_Agent_w_LLM/ollama_llm_client.py
# ollama_llm_client.py

import json
import requests
import os
from dotenv import load_dotenv

load_dotenv()
OLLAMA_URL = os.getenv("OLLAMA_URL", "http://localhost:11434")
MODEL_NAME = os.getenv("OLLAMA_MODEL", "llama3.2:3b")


class OllamaLLMClient:
    """LLM client for Ollama with tool-calling support."""

    def __init__(self, url: str = OLLAMA_URL, model: str = MODEL_NAME):
        self.url = url
        self.model = model

    def chat(self, messages, tools=None):
        """Send messages to Ollama and optionally receive tool calls."""
        prompt = self._build_prompt(messages, tools)
        try:
            resp = requests.post(
                f"{self.url}/api/generate",
                json={"model": self.model, "prompt": prompt, "stream": False, "temperature": 0.7},
                timeout=30
            )
            resp.raise_for_status()
            text = resp.json().get("response", "")
            return self._parse_response(text)
        except Exception as e:
            print(f"Ollama error: {e}")
            return self._mock_response(messages)

    def _build_prompt(self, messages, tools=None):
        parts = []
        for msg in messages:
            role = msg.get("role")
            content = msg.get("content")
            parts.append(f"{role.upper()}: {content}\n")
        if tools:
            parts.append("\nAvailable tools:\n")
            for t in tools:
                parts.append(f"- {t['name']}: {t.get('description', '')}\n")
        parts.append("\nASSISTANT: ")
        return "".join(parts)

    def _parse_response(self, text):
        """Try to parse JSON tool call from LLM."""
        try:
            start = text.find("{")
            end = text.rfind("}") + 1
            if start >= 0 and end > start:
                return json.loads(text[start:end])
        except:
            pass
        return {"response": text}

    def _mock_response(self, messages):
        user_msg = messages[-1]["content"]
        return {"response": f"LLM received: {user_msg}"}

File: 03_c_sharp/OllamaLLMClient.cs
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace MultiAgentFramework;

public class OllamaLLMClient
{
    private readonly string _url;
    private readonly string _model;
    private readonly HttpClient _http;

    public OllamaLLMClient(string? url = null, string? model = null)
    {
        _url   = url   ?? Environment.GetEnvironmentVariable("OLLAMA_URL")   ?? "http://localhost:11434";
        _model = model ?? Environment.GetEnvironmentVariable("OLLAMA_MODEL") ?? "llama3.2:3b";
        _http  = new HttpClient { Timeout = TimeSpan.FromSeconds(30) };
    }

    /// <summary>
    /// Send a list of chat messages to Ollama.
    /// Returns a parsed JsonObject if the response contains JSON, otherwise a plain text wrapper.
    /// </summary>
    public JsonObject Chat(List<ChatMessage> messages, List<ToolDefinition>? tools = null)
    {
        var prompt = BuildPrompt(messages, tools);

        try
        {
            var payload = JsonSerializer.Serialize(new
            {
                model       = _model,
                prompt      = prompt,
                stream      = false,
                temperature = 0.7,
            });

            var content  = new StringContent(payload, Encoding.UTF8, "application/json");
            var response = _http.PostAsync($"{_url}/api/generate", content).GetAwaiter().GetResult();
            response.EnsureSuccessStatusCode();

            var json = response.Content.ReadAsStringAsync().GetAwaiter().GetResult();
            var doc  = JsonDocument.Parse(json);
            var text = doc.RootElement.GetProperty("response").GetString() ?? "";

            return ParseResponse(text);
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Ollama error: {ex.Message}");
            return MockResponse(messages);
        }
    }

    // -------------------------------------------------------
    private string BuildPrompt(List<ChatMessage> messages, List<ToolDefinition>? tools)
    {
        var sb = new StringBuilder();

        foreach (var msg in messages)
            sb.AppendLine($"{msg.Role.ToUpper()}: {msg.Content}\n");

        if (tools != null && tools.Count > 0)
        {
            sb.AppendLine("\nAvailable tools:");
            foreach (var t in tools)
                sb.AppendLine($"- {t.Name}: {t.Description}");
        }

        sb.Append("\nASSISTANT: ");
        return sb.ToString();
    }

    private JsonObject ParseResponse(string text)
    {
        var start = text.IndexOf('{');
        var end   = text.LastIndexOf('}') + 1;

        if (start >= 0 && end > start)
        {
            try
            {
                var slice = text[start..end];
                var node  = JsonNode.Parse(slice);
                if (node is JsonObject obj) return obj;
            }
            catch { /* fall through */ }
        }

        return new JsonObject { ["response"] = text };
    }

    private JsonObject MockResponse(List<ChatMessage> messages)
    {
        var last = messages.LastOrDefault()?.Content ?? "";
        return new JsonObject { ["response"] = $"LLM received: {last}" };
    }
}

public record ChatMessage(string Role, string Content);
public record ToolDefinition(string Name, string Description);
